src/hsort_impl.h

```diff
@@ -0,0 +1,116 @@
+/***********************************************************************
+ * Copyright (c) 2021 Russell O'Connor, Jonas Nick                     *
+ * Distributed under the MIT software license, see the accompanying    *
+ * file COPYING or https://www.opensource.org/licenses/mit-license.php.*
+ ***********************************************************************/
+
+#ifndef SECP256K1_HSORT_IMPL_H
+#define SECP256K1_HSORT_IMPL_H
+
+#include "hsort.h"
+
+/* An array is a heap when, for all non-zero indexes i, the element at index i
+ * compares as less than or equal to the element at index parent(i) = (i-1)/2.
+ */
+
+static SECP256K1_INLINE size_t secp256k1_heap_child1(size_t i) {
+    VERIFY_CHECK(i <= (SIZE_MAX - 1)/2);
+    return 2*i + 1;
+}
+
+static SECP256K1_INLINE size_t secp256k1_heap_child2(size_t i) {
+    VERIFY_CHECK(i <= SIZE_MAX/2 - 1);
+    return secp256k1_heap_child1(i)+1;
+}
+
+static SECP256K1_INLINE void secp256k1_heap_swap64(unsigned char *a, size_t i, size_t j, size_t stride) {
+    unsigned char tmp[64];
+    VERIFY_CHECK(stride <= 64);
+    memcpy(tmp, a + i*stride, stride);
+    memmove(a + i*stride, a + j*stride, stride);
+    memcpy(a + j*stride, tmp, stride);
+}
+
+static SECP256K1_INLINE void secp256k1_heap_swap(unsigned char *a, size_t i, size_t j, size_t stride) {
+    while (64 < stride) {
+        secp256k1_heap_swap64(a + (stride - 64), i, j, 64);
+        stride -= 64;
+    }
+    secp256k1_heap_swap64(a, i, j, stride);
+}
```

Comment:
`secp256k1_heap_swap` is incorrect for `stride > 64`. See https://github.com/real-or-random/secp256k1/commit/95138ca6c1336bcd20ef332dd742f8a5f0f7cf75 for a fix and a test. This was overlooked because we never use such a stride, not even in the tests.

Alternatively, we could drop `secp256k1_heap_swap` entirely. `secp256k1_heap_swap64` appears to be correct. But I'd say let's just keep it if my fix looks good.

Reply:
FWIW, below is more like what I intended to write:

```c
static SECP256K1_INLINE void swap64(unsigned char *a, unsigned char *b, size_t len) {
    unsigned char tmp[64];
    VERIFY_CHECK(len <= 64);
    memcpy(tmp, a, len);
    memmove(a, b, len);
    memcpy(b, tmp, len);
}

static SECP256K1_INLINE void swap(unsigned char *arr, unsigned char *a, size_t stride) {
    unsigned char *a = arr + i*stride;
    unsigned char *b = arr + j*stride;
    size_t len = stride;
    while (64 < len) {
        swap64(a + (len - 64), b + (len - 64), 64);
        len -= 64;
    }
    swap64(a, b, len);
}
```

Obviously I've gotten this wrong before, so take it with a grain of salt.

Reply:
Looks correct (except that `i` and `j` should be in the parameter instead of `a`) to me, and is easier to read than my version, I think.

Reply:
Oops. Good catch @real-or-random. I cherry picked your test, copied @roconnor-blockstream's fix and improved the existing test.